gzip/bzip2 output options

b-wyss opened the pull request:
Feature creation in response to #505

mr-c commented:
Jenkins, retest this please

mr-c commented on setup.py:
```diff
@@ -15,6 +15,7 @@
 import shutil
 import subprocess
 import tempfile
+import bz2file as bz2
```
Not needed

mr-c commented:
Jenkins, test this please

mr-c commented:
Jenkins, test this please

mr-c commented:
@b-wyss Looks like you have some PEP8 violations. `make autopep8` should clear them up.

Keep on going, you've made good progress!

b-wyss commented:
Scripts to update:

- [x]normalize-by-median; 
- [ ]extract (which?); 
- [x]fasta-to-fastq; 
- [ ]filter-abund (single?); 
- [x]sample-reads-randomly; 
- [x]split paired-reads; 
- [x]interleave-reads;

ctb commented:
Looking at http://khmer.readthedocs.org/en/v1.3/user/scripts.html, I think:

filter-abund and filter-abund-single;
count-median;
extract-partitions;
extract-long-sequences;
extract-paired-reads;
fastq-to-fasta
interleave-reads
sample-reads-randomly
split-paired-reads

and eventually trim-low-abund when #759 is merged :)

ctb commented on khmer/kfile.py:
```diff
@@ -123,3 +125,17 @@ def check_valid_file_exists(in_files):
+                        help='Option to output as bz2')
+
+
+def enable_output_compression(args):
```
I think this can be generalized a bit - right now the output file has to be in args.output, but for some of the scripts I think there will be a need for different output names/files, especially in the scripts that have multiple output files.

So maybe enable_output_compression can take a file handle, and return an fp (what gzip.GzipFile/bz2file.open give) for it?

ctb replied:
(note: open for discussion. just a thought.)

ctb commented on scripts/normalize-by-median.py:
```diff
@@ -204,12 +206,14 @@ def get_parser():
 def main():  # pylint: disable=too-many-branches,too-many-statements
     info('normalize-by-median.py', ['diginorm'])
     args = get_parser().parse_args()
+    enable_output_compression(args)
```
see, here it will only work if you specify a -o, which is good, but it'd be nice if this worked for any and all output files no?

mr-c commented:
@ctb First pass is for single outputs ( `-o` or similar)

bocajnotnef commented:
@b-wyss Mind if I vulture this?

mr-c commented:
Go for it

On Tue, Jul 21, 2015 at 10:52 AM Jake Fenton <notifications@github.com>
wrote:

> @b-wyss <https://github.com/b-wyss> Mind if I vulture this?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dib-lab/khmer/pull/747#issuecomment-123414718>.
>
-- 
Michael R. Crusoe: Programmer & Bioinformatician crusoe@ucdavis.edu
The lab for Data Intensive Biology; University of California, Davis
https://impactstory.org/MichaelRCrusoe http://twitter.com/biocrusoe

b-wyss commented:
Please do! Sorry I fell off the face of the earth, things got really crazy
towards the end of the spring, centered around a bit of a family situation.
I meant to officially leave and try to make it on good terms, but it never
really happened.  You guys deserved to at least know what was going on with
me, and I didn't let anyone know.  Seems like the lab has transferred over
pretty successfully - congratulations!

On Tue, Jul 21, 2015 at 1:57 PM Michael R. Crusoe <notifications@github.com>
wrote:

> Go for it
>
> On Tue, Jul 21, 2015 at 10:52 AM Jake Fenton <notifications@github.com>
> wrote:
>
> > @b-wyss <https://github.com/b-wyss> Mind if I vulture this?
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/dib-lab/khmer/pull/747#issuecomment-123414718>.
> >
> --
> Michael R. Crusoe: Programmer & Bioinformatician crusoe@ucdavis.edu
> The lab for Data Intensive Biology; University of California, Davis
> https://impactstory.org/MichaelRCrusoe http://twitter.com/biocrusoe
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dib-lab/khmer/pull/747#issuecomment-123415854>.
>

bocajnotnef commented:
All good! Stuff happens. Hope everything is okay and you're doing well!

bocajnotnef commented:
TO-DO:

 - [ ] Generalize `enable compression`: currently it takes `args` and converts `args.output` to a gz/bz writer. as per @ctb reccomendation, will shift to a func that takes a file handle and returns an FP. This will enable use in scripts a la normalize-by-median with multiple files.
 - [ ] Enable compression in normalize-by-median
   - [ ] determine which outputs in normalize-by-median should be compressed

bocajnotnef commented:
question: should the func that adds the compression args be moved to khmer_args?

https://github.com/dib-lab/khmer/pull/747/files#diff-46d242f8c7f4ca46fd5c784bd24184c7R165

bocajnotnef commented:
normalize-by-median's stdout 3 test is leaking output

bocajnotnef commented:
"revert" to argparse handling opening files:

- [x] normalize-by-median (multiple files); 
- [x] count-median;
- [x] extract-partitions; 
- [x] extract-long-sequences;
- [x] extract-paired-reads;
- [x] fastq-to-fasta; 
- [x] filter-abund;
- [x] filter-abund-single; 
- [x] sample-reads-randomly; 
- [x] split paired-reads; 
- [x] interleave-reads;
- [x] trim-low-abund;

bocajnotnef commented:
`split-paired-reads` has this:

``` parser.add_argument('infile', nargs='?', default='/dev/stdin') ```

we should make than an argparse handled file open

bocajnotnef commented on khmer/kfile.py:
```diff
@@ -167,3 +169,31 @@ def check_valid_file_exists(in_files):
+    if file_handle is sys.stdout:
+        return sys.stdout
+    else:
+        assert type(file_handle) == file, type(file_handle)
```
this is primarily a debugging thing and will be removed

ctb replied:
*shrug* it's ok to leave it in, with a brief explanation.

bocajnotnef commented:
retest this, please

ctb commented on scripts/fastq-to-fasta.py:
```diff
@@ -67,7 +70,7 @@ def main():
         print('No lines dropped from file.', file=sys.stderr)
 
-    print('Wrote output to', args.output, file=sys.stderr)
+    print('Wrote output to', str(args.output), file=sys.stderr)
```
I think this can be args.output.name, no?

bocajnotnef replied:
Fails if we pass sys.stdout as output--it doesn't have a name.

That said, my `str` solution doesn't really help. I should check to see if it is in fact a block device or whathaveyou.

@mr-c reccomended making a function to do checking on file handles to see if they are fifos, block devices, etc. and factoring those checks out of the rest of the codebase.

ctb replied:
> @mr-c reccomended making a function to do checking on file handles to see if they are fifos, block devices, etc. and factoring those checks out of the rest of the codebase.

yarp. +1.

ctb commented on khmer/kfile.py:
```diff
@@ -167,3 +169,31 @@ def check_valid_file_exists(in_files):
+def get_file_writer(file_handle, do_gzip, do_bzip):
+    ofile = None
+
+    if do_gzip and do_bzip:
```
I think this can be handled at the option parsing stage by using mutually exclusive groups:

https://docs.python.org/2/library/argparse.html#mutual-exclusion

bocajnotnef commented:
TO-DO:

- [ ] fix jenkins failures: `AttributeError: 'FileType' object has no attribute '__name__'`--test failure indicator thing helpfully me to add 'metavar=filetype' to all the filetype invocations in argparse (wooo)
- [ ] add func to detect block devices
- [ ] I broke writing compressed data to stdout when I wasn't looking--gotta fix that.
- [ ] diff-cover, etc (oh joy)

bocajnotnef commented:
fun little tiddlybit: bz2file doesn't give you a way to get the path from an open file object, so if you're gonna run stats on things (to, oh, I dunno, see if they're block devices) you have to do that before nay sort of compression handling happens.

bocajnotnef commented:
coevrage gap: `scripts/split-paired-reads.py (85.7%): Missing lines 107`
https://github.com/dib-lab/khmer/pull/747/files#diff-d6c1b656418a05c32a5783ca39ed52ebR107
Non-critical error output that's not fun to make a test case to reproduce--I think it's fine to not test.

bocajnotnef commented:
retest this, please

bocajnotnef commented:
question: should I move the func that adds compression args into khmer_args?

https://github.com/dib-lab/khmer/pull/747/files#diff-46d242f8c7f4ca46fd5c784bd24184c7R184

bocajnotnef commented on tests/test_scripts.py:
```diff
@@ -2354,6 +2389,52 @@ def test_sample_reads_randomly_force_single():
     assert seqs == answer
 
 
+def test_sample_reads_randomly_force_single_outfile():
```
I didn't write this and as such have no idea why it's here.

bocajnotnef commented:
retest this please

bocajnotnef commented:
- [x] Is it mergeable?
- [x] Did it pass the tests?
- [x] If it introduces new functionality in scripts/ is it tested?
  Check for code coverage with `make clean diff-cover`
- [x] Is it well formatted? Look at `make pep8`, `make diff_pylint_report`,
  `make cppcheck`, and `make doc` output. Use `make format` and manual
  fixing as needed.
- [x] Did it change the command-line interface? Only additions are allowed
  without a major version increment. Changing file formats also requires a
  major version number increment.
- [x] Is it documented in the ChangeLog?
  http://en.wikipedia.org/wiki/Changelog#Format
- [x] Was a spellchecker run on the source code and documentation after
  changes were made?
- [x] Is the Copyright year up to date?

@ctb @mr-c CR/Merge, please?

mr-c commented on khmer/kfile.py:
```diff
@@ -167,3 +169,45 @@ def check_valid_file_exists(in_files):
+        return True
+    else:
+        mode = os.stat(fthing.name).st_mode
+        return S_ISBLK(mode) or S_ISCHR(mode)
```
`or S_ISFIFO(mode)`

bocajnotnef replied:
The primary use case I see for this function is for something governing output; i.e.
```
if is_block()
 print ("wrote to block thing")
else:
 print("wrote to" + filepath)
```
fifos have paths and thus aren't a concern in this case.

mr-c replied:
Or file size checks

bocajnotnef replied:
Hm?

mr-c replied:
I would assume if I was doing a file size check that I could use `is_block` to skip some files.

bocajnotnef replied:
Yes--but not `S_ISFIFO(mode)`

mr-c commented on khmer/kfile.py:
```diff
@@ -167,3 +169,45 @@ def check_valid_file_exists(in_files):
+    """Adds compression arguments to a parset object"""
+    group = parser.add_mutually_exclusive_group()
+    group.add_argument('--gzip', default=False, action='store_true',
+                       help='Option to output as gzip')
```
`help='Compress output using gzip')`

mr-c commented on khmer/kfile.py:
```diff
@@ -167,3 +169,45 @@ def check_valid_file_exists(in_files):
+    group.add_argument('--gzip', default=False, action='store_true',
+                       help='Option to output as gzip')
+    group.add_argument('--bzip', default=False, action='store_true',
+                       help='Option to output as bz2')
```
help='Compress output using bzip2')

mr-c commented on khmer/kfile.py:
```diff
@@ -167,3 +169,45 @@ def check_valid_file_exists(in_files):
+    elif do_bzip:
+        ofile = bz2file.open(file_handle, mode='w')
+    else:
+        return file_handle
```
I would say `ofile = file_handle` so that there is just one `return`

mr-c commented on scripts/extract-paired-reads.py:
```diff
@@ -68,15 +70,16 @@ def get_parser():
                         'directory if necessary')
 
-    parser.add_argument('-p', '--output-paired', metavar='output_paired',
+    parser.add_argument('-p', '--output-paired', metavar="filename",
```
Long argument name first, please

mr-c commented on scripts/extract-partitions.py:
```diff
@@ -214,7 +217,9 @@ def main():  # pylint: disable=too-many-locals,too-many-branches
     group_fps = {}
     for _ in range(group_n):
-        group_fp = open('%s.group%04d.%s' % (args.prefix, _, suffix), 'w')
+        group_fp = get_file_writer(open('%s.group%04d.%s' %
```
What a messy line! :-) Care to decompose this for readability?

mr-c commented on scripts/fastq-to-fasta.py:
```diff
@@ -67,7 +72,10 @@ def main():
 
-    print('Wrote output to', args.output, file=sys.stderr)
+    if output_is_block:
+        print('Wrote output to block device', file=sys.stderr)
```
What's wrong with `args.output.name` ?

bocajnotnef replied:
Doesn't work on `sys,stdout` and friends.

bocajnotnef replied:
Strike that, it kinda works

```
In [1]: import sys

In [2]: f = open(sys.stdout, 'w')
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-3a6ff793e3af> in <module>()
----> 1 f = open(sys.stdout, 'w')

TypeError: coercing to Unicode: need string or buffer, file found

In [3]: sys.stdout.name
Out[3]: '<stdout>'
```

mr-c commented on scripts/interleave-reads.py:
```diff
@@ -59,11 +59,12 @@ def get_parser():
     parser.add_argument('-o', '--output', metavar="filename",
                         type=argparse.FileType('w'),
-                        default=sys.stdout)
+                        default=None)
```
Other scripts have defaults of `sys.stdout` -- why not here?

bocajnotnef replied:
It messes with detecting if the user gave a name--normally we'd just check to see if `args.output is not none` or summuch, but if we make it standard to sys.stdout it always looks like `-o` has been specified.

In interleave- if we don't specify `-o` the script automatically comes up with its own names.

mr-c commented on scripts/normalize-by-median.py:
```diff
@@ -257,15 +260,16 @@ def get_parser():
-                        dest='single_output_file',
                         type=argparse.FileType('w'),
-                        default=None, help='only output a single file with '
+                        dest='single_output_file',
```
Why was the default eliminated?

mr-c replied:
See also https://github.com/dib-lab/khmer/pull/1186/files#r35375513

mr-c commented on scripts/sample-reads-randomly.py:
```diff
@@ -162,8 +163,8 @@ def main():
               (len(reads[0]), output_filename), file=sys.stderr)
-        if not output_file:
-            output_file = open(output_filename, 'w')
+        output_file = get_file_writer(open(output_filename, 'w'), args.gzip,
```
I usually put a newline after the first open parenthesis to make the long lines like this wrap better

mr-c commented on scripts/split-paired-reads.py:
```diff
@@ -72,11 +73,13 @@ def get_parser():
     parser.add_argument('-1', '--output-first', metavar='output_first',
+                        type=argparse.FileType('w'),
                         default=None, help='Output "left" reads to this '
-                        'file', type=argparse.FileType('w'))
```
You have the noisiest diffs! Half of these changes make no difference; what's up with that? :-)

bocajnotnef replied:
What's weird is I swore I added that back in.

bocajnotnef replied:
Oh, I know what happened. I took it out and added it back it, but just flipped where it was.

Whoops.

mr-c commented on tests/test_scripts.py:
```diff
@@ -1626,6 +1626,41 @@ def test_do_partition_2_fq():
+    # sys.stdout = capture = StringIO
+
+    script = 'interleave-reads.py'
+    args = [infile1, infile2]
```
No compression options?

bocajnotnef replied:
Not here, no--I test compressed to stdout in one of the fastq-to-fasta tests

SensibleSalmon committed:
Added output compression and associated tests

bocajnotnef commented:
retest this please

bocajnotnef commented:
@mr-c Updated and responses to comments

SensibleSalmon committed:
made load-graph no longer puke .pt's all over its output filenames
Changed tests and other scripts accordingly

SensibleSalmon committed:
changelog update

bocajnotnef commented:
gotta do a pep257

SensibleSalmon committed:
pep257

bocajnotnef commented:
retest this please

SensibleSalmon committed:
Merge branch 'master' of github.com:ged-lab/khmer into feature/gzip505

Conflicts:
	ChangeLog

SensibleSalmon committed:
cleanup

SensibleSalmon committed:
we do need fifos, surprise.

bocajnotnef commented:
@mr-c CR, please?

bocajnotnef commented:
- [ ] Is it mergeable?
- [ ] Did it pass the tests?
- [ ] If it introduces new functionality in scripts/ is it tested?
  Check for code coverage with `make clean diff-cover`
- [ ] Is it well formatted? Look at `make pep8`, `make diff_pylint_report`,
  `make cppcheck`, and `make doc` output. Use `make format` and manual
  fixing as needed.
- [ ] Did it change the command-line interface? Only additions are allowed
  without a major version increment. Changing file formats also requires a
  major version number increment.
- [ ] Is it documented in the ChangeLog?
  http://en.wikipedia.org/wiki/Changelog#Format
- [ ] Was a spellchecker run on the source code and documentation after
  changes were made?
- [ ] Is the Copyright year up to date?

mr-c commented:
No CR without a completed checklist :-)

SensibleSalmon committed:
Merge branch 'master' of github.com:ged-lab/khmer into feature/gzip505

Conflicts:
	ChangeLog
	scripts/count-median.py
	tests/test_scripts.py

bocajnotnef commented:
- [ ] Is it mergeable?
- [ ] Did it pass the tests?
- [ ] If it introduces new functionality in scripts/ is it tested?
  Check for code coverage with `make clean diff-cover`
- [ ] Is it well formatted? Look at `make pep8`, `make diff_pylint_report`,
  `make cppcheck`, and `make doc` output. Use `make format` and manual
  fixing as needed.
- [ ] Did it change the command-line interface? Only additions are allowed
  without a major version increment. Changing file formats also requires a
  major version number increment.
- [ ] Is it documented in the ChangeLog?
  http://en.wikipedia.org/wiki/Changelog#Format
- [ ] Was a spellchecker run on the source code and documentation after
  changes were made?
- [ ] Is the Copyright year up to date?

SensibleSalmon committed:
cleanup

mr-c commented:
The what's new file will need to be updated with the `.pt` business.

SensibleSalmon committed:
updated whatsnew for load/partition-graph

ctb commented on khmer/kfile.py:
```diff
@@ -167,3 +168,43 @@ def check_valid_file_exists(in_files):
+    if is_block(file_handle):
+        pass
+    else:
+        assert type(file_handle) == file, type(file_handle)  # Sanity check
```
This is actually unpythonic - see Duck Typing :).  Basically, we're supposed to test behavior, not types.

bocajnotnef replied:
Alright. Should I bother with sanity checking at all, then?

ctb replied:
> Alright. Should I bother with sanity checking at all, then?

Presumably any mistakes will cause problems on the first actual `write` call,
yes?  So then not needed.

Also note that 'assert' is, in any case, removed when running with
optimization.  So it's best used for sanity checking in developer and
testing mode, not for actual runtime sanity checking.

ctb replied:
On Tue, Jul 28, 2015 at 10:25:29AM -0700, Jake Fenton wrote:
> > +    if is_block(file_handle):
> > +        pass
> > +    else:
> > +        assert type(file_handle) == file, type(file_handle)  # Sanity check

also for future I would say you should write this as

```
if not is_block(file_handle): ...
```

SensibleSalmon committed:
pythonicism is good

bocajnotnef commented:
@mr-c @ctb Cleaned up, ready for merge.

SensibleSalmon committed:
Merge branch 'master' of github.com:ged-lab/khmer into feature/gzip505

Conflicts:
	ChangeLog

bocajnotnef commented:
retest this please

SensibleSalmon committed:
Merge branch 'master' of github.com:ged-lab/khmer into feature/gzip505

Conflicts:
	ChangeLog

mr-c commented on ChangeLog:
```diff
@@ -1,3 +1,14 @@
+2015-07-28  Jacob Fenton  <bocajnotnef@gmail.com>
+
+   * khmer/kfile.py: added infastructure for doing compressed output
```
Did you run codespell?

mr-c replied:
Ah, `codespell` only catches common misspellings. People should still use aspell or other checker.

mr-c commented on ChangeLog:
```diff
@@ -1,3 +1,14 @@
+   * tests/test_{functions,scripts}.py: added tests
+   * scripts/{load-graph,partition-graph}.py: made load-graph no longer add .pt
+   to graph outfiles, changed partition-graph to not expect .pt's
+
```
at least one file is missing from this ChangeLog

SensibleSalmon committed:
misspelling

SensibleSalmon committed:
Merge branch 'master' of github.com:ged-lab/khmer into feature/gzip505

Conflicts:
	ChangeLog
	khmer/kfile.py
	scripts/count-median.py
	scripts/extract-long-sequences.py
	scripts/fastq-to-fasta.py
	scripts/filter-abund.py
	scripts/interleave-reads.py
	scripts/normalize-by-median.py
	scripts/split-paired-reads.py

SensibleSalmon committed:
fixes and cleanup from the merge

SensibleSalmon committed:
Merge branch 'master' of github.com:ged-lab/khmer into feature/gzip505

Conflicts:
	ChangeLog

bocajnotnef commented:
retest this please

bocajnotnef commented:
Huzzah, tests passing.

@mr-c @luizirber CR, please?

mr-c commented on khmer/kfile.py:
```diff
@@ -170,3 +171,38 @@ def check_valid_file_exists(in_files):
+
+
+def add_output_compression_type(parser):
+    """Add compression arguments to a parset object."""
```
parset?

mr-c commented:
The tests fail on Python3

bocajnotnef commented:
Things we learned today:

 - bz2file needs to be given a binary file handle to play nice with python3
 - all our problems are solved by telling argparse to open output files in binary mode and then using `write_record` to output records (which we should be doing anyway, yes?)
 - we need moar test coverage for this PR--technically everything is tested, but we don't test all the scripts to see if compression works with them.
 
Additionally, I need to sit down with someone and see if the outputs I am compressing are outputs that should be compressed--that is, I currently have the csv output from count-median compressed, which seems a bit silly.

For the files I've changed;
```
script                  output type     should be   is      uses write_record
                                        compressed

count-median            csv report      no          yes     N/Aish
extract-long-seq        sequences       yes         yes     yes
extract-paired-reads    sequences       yes         yes     yes
extract-partitions      sequences       yes         yes     yes
fastq-to-fasta          sequences       yes         yes     does now

filter-abund-single     sequences       yes         yes     no

        note: uses threaded_sequence_processor from thread_utils, which does
        NOT use write_record and probably totally should
        
                        counting table  ???         no      --

filter-abund            sequences       yes         yes     no
interleave-reads        sequences       yes         yes     yes
normalize-by-median     sequences       yes         yes     yes
                        hash(?)table    ???         no      --
sample-reads-randomly   sequences       yes         yes     yes
split-paired-reads      sequences       yes         yes     yes
trim-low-abund          sequences       yes         yes     yes
                        paired seq      yes         NO      yes     !!!
                        hash(?)table    ???         no      --
```

So, next up: I fix trim-low to compress paired seqs, I make count-median not compress a report, and people that understand the software base better than I do help sort out the make-all-argparse-file-opens-binary-and-make-things-use-write-record-where-they-should thing.

And maybe a crapton of tests.

ctb commented:
> Additionally, I need to sit down with someone and see if the outputs I am compressing are outputs that should be compressed--that is, I currently have the csv output from count-median compressed, which seems a bit silly.

Good catch.

Only write_record output should be compressed.

bocajnotnef commented:
Additional note; not all outfiles are opened by argparse (extract-paired, extract-partition, filter-abund-single, filter-abund, normalize, sample-randomly, split-paired, trim-low)--primarily when the outfile name is auto'd if a single one isn't specified.

So. Those need to be set to binary.

bocajnotnef commented:
@ctb That was @mr-c'c catch

bocajnotnef commented:
TO-DO:

- [ ] fix trim-low to compress paired seqs
- [ ] make count-median not compress a report, 
- [ ] issuefy making filter-abund*'s threaded_sequence_processor use write_record
- [ ] tests?
- [ ] open errythang in binary mode and throw it at write_record
- [ ] make sure we're only compressing things thrown at write_record

SensibleSalmon committed:
@luizirber py3 fixes, spelling

SensibleSalmon committed:
pep8

SensibleSalmon committed:
made threaded_sequence_processor use write_record
made count-median no longer compress reports
made trim-low-abund open a thing in binary (one down, 1238947 to go)

SensibleSalmon committed:
Merge branch 'master' of github.com:ged-lab/khmer into feature/gzip505

Conflicts:
	ChangeLog

bocajnotnef commented:
question re trim-low-abund: we set aside stuff in a file, then go back to that file, pull things from it and *then* write to output. Currently, I don't compress the aside_file--Should I?

SensibleSalmon committed:
write everything to binary; resist urge to make commit message 01100010 01101001 01101110 01100001 01110010 01111001

ctb commented:
Good question. No for now, but file an issue.

--
Titus Brown, ctbrown@ucdavis.edu

> On Jul 30, 2015, at 11:22 AM, Jake Fenton <notifications@github.com> wrote:
> 
> question re trim-low-abund: we set aside stuff in a file, then go back to that file, pull things from it and then write to output. Currently, I don't compress the aside_file--Should I?
> 
> —
> Reply to this email directly or view it on GitHub.
>

bocajnotnef commented:
Right. Everything is in place except for more extensive test coverage.

Probably gonna make a `test_output_compression.py` or summuch

ctb commented:
I would suggest adding streaming tests to the new test_streaming_io.

--
Titus Brown, ctbrown@ucdavis.edu

> On Jul 30, 2015, at 11:31 AM, Jake Fenton <notifications@github.com> wrote:
> 
> Right. Everything is in place except for more extensive test coverage.
> 
> Probably gonna make a test_output_compression.py or summuch
> 
> —
> Reply to this email directly or view it on GitHub.
>

bocajnotnef commented:
Agreed, but my concern is that for nearly every case where we have an output file we could have a compressed output file, and we should be testing all of those--but that's like, 50 tests.

bocajnotnef commented:
from high-bandwidth conversation; We'll employ stupidity-driven-testing here. No point in going through the combinatorial matrix (Well, there is a point, but the cost/benefit isn't there). So if somebody finds something that's super borked we'll fix it and add a test for it but as we stand we're probably on solid footing.

bocajnotnef commented:
On that note, @ctb Merge?

ctb commented on scripts/fastq-to-fasta.py:
```diff
@@ -55,8 +61,8 @@ def main():
 
-        args.output.write('>' + name + '\n')
-        args.output.write(sequence + '\n')
+        del record['quality']
```
lol, ok ;)

ctb commented on scripts/sample-reads-randomly.py:
```diff
@@ -166,8 +167,13 @@ def main():
-        if not output_file:
-            output_file = open(output_filename, 'w')
+
+        if args.output_file:
```
Change this to do one call to get_file_writer, pls.

```
output_file = args.output_file
if not output_file: ...
outputf_file = get_file_writer(...)
```

ctb commented on setup.py:
```diff
@@ -178,9 +178,7 @@ def check_for_openmp():
-        "install_requires": ['screed >= 0.9'],
-        # testing screed download link
-
+        "install_requires": ['screed >= 0.9', 'bz2file'],
```
what happens if bz2file isn't available? I'm -0 on making this a hard requirement (although it may come with all the Pythons now).

mr-c replied:
`bz2file` is already a requirement of screed

On Fri, Jul 31, 2015 at 7:23 AM C. Titus Brown <notifications@github.com>
wrote:

> In setup.py
> <https://github.com/dib-lab/khmer/pull/747#discussion_r35977820>:
>
> > @@ -178,9 +178,7 @@ def check_for_openmp():
> >          "url": 'https://khmer.readthedocs.org/',
> >          "packages": ['khmer', 'khmer.tests', 'oxli'],
> >          "package_dir": {'khmer.tests': 'tests'},
> > -        "install_requires": ['screed >= 0.9'],
> > -        # testing screed download link
> > -
> > +        "install_requires": ['screed >= 0.9', 'bz2file'],
>
> what happens if bz2file isn't available? I'm -0 on making this a hard
> requirement (although it may come with all the Pythons now).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dib-lab/khmer/pull/747/files#r35977820>.
>
-- 
Michael R. Crusoe: Programmer & Bioinformatician crusoe@ucdavis.edu
The lab for Data Intensive Biology; University of California, Davis
https://impactstory.org/MichaelRCrusoe http://twitter.com/biocrusoe

ctb replied:
On Fri, Jul 31, 2015 at 07:25:26AM -0700, Michael R. Crusoe wrote:
> > @@ -178,9 +178,7 @@ def check_for_openmp():
> >          "url": 'https://khmer.readthedocs.org/',
> >          "packages": ['khmer', 'khmer.tests', 'oxli'],
> >          "package_dir": {'khmer.tests': 'tests'},
> > -        "install_requires": ['screed >= 0.9'],
> > -        # testing screed download link
> > -
> > +        "install_requires": ['screed >= 0.9', 'bz2file'],
> 
> `bz2file` is already a requirement of screed

ok, same comment applies there :). @bocajnotnef punt to an issue if
you want.

bocajnotnef replied:
#1217

ctb commented on tests/test_functions.py:
```diff
@@ -27,6 +27,17 @@ def test_forward_hash():
+    try:
+        get_file_writer(somefile, True, True)
+        assert False, "Shouldn't get to this"
+    except Exception as err:
```
AssertionError is a subclass of exception, so you need to take a different approach here.  I would suggest setting a flag in place of the first assert, and then checking to make sure the flag *isn't* set.

ctb commented:
* The number of changed files and the number of files mentioned in the ChangeLog entry are different;
   what's up?
* Is `is_block` really only used in one place in the codebase? Is there nowhere else it belongs?
* More generally, in the 'Wrote output to" section of fastq-to-fasta - is there a need for a general way to
   provide a human-readable output name for things that may-or-may-not-be-files? I would prefer that
   logic like `if output_is_block` be placed somewhere central and reusable (kfile.py?)  - perhaps a
   function named 'describe_filehandle` or some such that returns a string?

ctb commented:
Overall looks good. Double-check your diff-cover, fix issues above, ask for re-review :)

SensibleSalmon committed:
cleanup, minor refactor for generalizing output info logic

SensibleSalmon committed:
Merge branch 'master' of github.com:ged-lab/khmer into feature/gzip505

Conflicts:
	ChangeLog

bocajnotnef commented:
@ctb Cleaned up, ready for merge.

ctb committed:
Merge branch 'master' of github.com:dib-lab/khmer into feature/gzip505

Conflicts:
	ChangeLog

ctb committed:
minor cleanup

ctb committed:
updated ChangeLog; preserve minimum greppability

ctb committed:
more minor cleanup

ctb committed:
fix heading foo

ctb committed:
updated ChangeLog

ctb commented:
LGTM.

bocajnotnef commented:
Thanks!
